### PR TITLE
Add gen8 support

### PR DESCRIPTION
**Pull request by Yoda2798**
Added support for generation 8 (SS) Pokemon, abilities, items, moves, learnsets by updating with files from https://github.com/smogon/pokemon-showdown/tree/master/data.

Added sets from the gen 8 file in https://github.com/smogon/damage-calc/tree/master/src/js/data/sets (which is mainly old gen 7 ones right now) for use in SS.

Tweaked js and html files so that SS is an available, and now the default, generation.

image

### CODE REVIEW

**Comment by jaredgrady** [comment]
Thanks! This looks great. I’ll do some validation and merge and deploy it to prod this week.

**Comment by jaredgrady** [comment]
One thought. We may need to change the default team. I don’t think unown is available in gen 8. Might be better to change it to something else. Thoughts @Nineage?

**Comment by Yoda2798** [comment]
Unown isn't available right now, but IMO it still stylistically fits as the default Pokemon. I think there is merit in keeping 'snapped' Pokemon available for people building in natdex tiers. What I would bring up from that, however, is the issue of the builder not doing any sort of legality check on pokemon/moves/etc. like you can put any move or ability on any Pokemon, or add megas and Tapus in DPP. That seemed more complicated to change though so I haven't looked into it.

**Comment by jaredgrady** [comment]
We talked about team validation but it is definitely a complex topic. We’d have to be on top of maintaining the rules for various metagames. We could implement simple logic, such as not allowing mons from a later generation in an early one, etc. Let me ask around and see if anyone maintains an api or module that can make that easy for us.

**Comment by Yoda2798** [comment]
Yeah I think something simple like not allowing pokemon from later generations, and only showing moves/abilities learnable by the given Pokemon (and only those available in gen 8 for SS; but in/before given gen for older ones if possible) would go a long way. For Pokemon available in that generation, going by dex no. and -alola, -galar, -gmax, for the exceptions should work. Abilities for each Pokemon are in the pokedex, and moves are in learnsets (with number at front e.g. 5 in "5L1" representing the generation it's learnt in). If you don't hear anything back I might look into it at some point in the future.

**Comment by jaredgrady** [comment]
Do you think there is anything on here we could use for images of the new Pokemon [https://pldh.net/media/pokemon/](https://pldh.net/media/pokemon/). It would be nice to still have some images for them even if they aren't the shuffle images

**Comment by Yoda2798** [comment]
There's (the official artwork) images for all Pokemon in https://pldh.net/media/pokemon/sugimori/ which could work, image previewing how it would look below. On that point, the collection of images you're currently using from that site is missing some Pokemon, such as Kartana and Mega Pokemon, which Bulbapedia does have (https://archives.bulbagarden.net/w/index.php?title=Category:Shuffle_icons&fileuntil=115%0AShuffle115M.png#mw-category-media). Image below also shows the difference between current icons and using the additional ones from Bulbapedia.

teambuilder3

**Comment by jaredgrady** [comment]
Yeah I like how that looks. I can work on implementing that unless it’s something you feel comfortable doing

**Comment by Yoda2798** [comment]
I just made a commit using the Sugimori artwork for gen 8 Pokemon (and the USUM ultrabeasts like Naganadel which also don't have shuffle images) and taking sprites from https://www.pkparaiso.com/shuffle/sprites.php (since the urls were simpler than from Bulbapedia) for the other few Pokemon like Kartana. Correct images aren't used yet for mega/galar/alola Pokemon but I'm looking into that now.

**Comment by Yoda2798** [comment]
Okay, mega/galar/alola Pokemon work properly now. For Gmax Pokemon the image for the regular version is used, e.g. for Charizard it'll be the shuffle icon, and for Hatterene it'll be the Sugimori artwork. I got a bit carried away and decided to add basically all of the forme changes with a shuffle icon, whoops. Stuff with alternate icons that I didn't add were gendered Wobuffet, Jellicent, and Meowstic; shinies; and different formes for Castform, Burmy, Shellos/Gastrodon, Deerling/Sawsbuck, Vivillon, and Pikachu. Aside from those, I added the ones for like, everything, from primals to Rotom and Deoxys to Keldeo, Shaymin, and Giratina. 

If you want to see all the forme differences working here's some pictures (these would all have had the same icon before):
https://puu.sh/F0jGK/c8b149bdfc.png 
https://puu.sh/F0jH4/af8a3a5a52.png 
https://puu.sh/F0jHH/5f2488b2c2.png 
https://puu.sh/F0jIi/c1c217f85a.png 
https://puu.sh/F0jIV/b4d5748861.png 
https://puu.sh/F0jJV/844b0684b4.png 
https://puu.sh/F0jK6/cc60efafba.png 
https://puu.sh/F0jKu/acfa7e9c34.png 
https://puu.sh/F0jWM/fda0223291.png

teambuilder4

**Comment by jaredgrady** [comment]
Sweeeeet. Looks good to me. I'm just waiting for nineage to update the API key so i can deploy this using CI/CD

**Comment by Yoda2798** [comment]
I decided to go about doing legality stuff; when you search for Pokemon and items it will only show results from that generation or earlier (with megas/gmax/galar/alola formes only in the correct generations). Also for gen 8 everything is shown, basically assuming NatDex format, rather than what's currently available in SwSh. Abilities shown are the ones that the given Pokemon gets, and if the Pokemon is Mega, then also the abilities of the regular forme. Hidden Ability is not shown in DPP. Moves shown are ones learnable by the given Pokemon, or by the regular forme/pre-evolutions, by the current generation. For example, Rotom-Wash will also use regular Rotom's moves (each forme has only the special move in learnsets), and Mega Venusaur will combine the movepools for Bulbasaur, Ivysaur, and Venusaur. I don't know if there's anything missing doing that or if there was an easier way copying from the PS code or something but it seems to work. The code is fairly messy though, so there could well be a mistake.

Gothitelle Fake Out in SS: http://puu.sh/F0ErR/92c50363d4.png vs in SM: http://puu.sh/F0Esq/0c2a5c03fd.png
Mega Venusaur abilities: http://puu.sh/F0F8b/13dcaa2fcb.png
Charizard formes in SS: http://puu.sh/F0Fc9/d8a303c4ec.png vs in XY: http://puu.sh/F0FdH/9643f4ae5e.png vs in BW: http://puu.sh/F0FeK/9976d8ecab.png
Heavy-Duty Boots in SS: http://puu.sh/F0Fi3/67ad60b6b9.png vs in SM: http://puu.sh/F0FiP/1c1a556b8a.png

File: public/js/events.js [hunk 1]
```diff
@@ -197,28 +204,108 @@ function getImportable() {
     return output.join("\n");
 }
 
+// get legal moves for a Pokemon (forme)
+function getMoves(moves_in, pokemon, learnsets) {
+	if (!learnsets[pokemon]) return moves_in;
+	var genNo = 8;
+	if (teamGen == "SM") genNo = 7;
```
Comment (jaredgrady) [inline public/js/events.js]: Can we use `===` instead for these. Not a huge fan of type coercion.

File: public/js/events.js [hunk 2]
```diff
@@ -197,28 +204,108 @@ function getImportable() {
...
+		dex = Object.keys(dataFile[monName]['abilities']);
+		for (var i = 0; i < dex.length; i++) {
+			if (dex[i] == 'H' && teamGen == "DPP") continue;
+			data.push(dataFile[monName]['abilities'][dex[i]]);
+		}
+		if (['Mega','Mega-X','Mega-Y'].includes(pokedex[monName]['forme'])) {
+			monName = toId(pokedex[monName]['baseSpecies']);
+			dex = Object.keys(dataFile[monName]['abilities']);
+			for (var i = 0; i < dex.length; i++) {
+				if (dex[i] == 'H' && teamGen == "DPP") continue;
```
Comment (jaredgrady) [inline public/js/events.js]: Here as well with the strict equals operator

File: public/js/events.js [hunk 3]
```diff
@@ -197,28 +204,108 @@ function getImportable() {
...
+	        if (dexNo > 494 && teamGen == "DPP") continue;
+			if (dexNo > 649 && teamGen == "BW") continue;
+			if (dexNo > 721 && teamGen == "XY") continue;
+			if (dexNo > 809 && teamGen == "SM") continue;
+			if (['Totem','Busted','Busted-Totem'].includes(forme) && teamGen !== "SS") continue;
+			if (['Mega','Mega-X','Mega-Y'].includes(forme) && ['DPP','BW'].includes(teamGen)) continue;
+			if (forme == 'Alola' && ['DPP','BW','XY'].includes(teamGen)) continue;
+			if (['Gmax','Galar','Galar-Zen'].includes(forme) && teamGen !== "SS") continue;
+			if (forme == "Spiky-eared" && teamGen !== "DPP") continue;
+			if (forme == "Starter" && teamGen !== "SM") continue;
```
Comment (jaredgrady) [inline public/js/events.js]: and here

**Comment by Yoda2798** [comment]
Done, changed them all.

**Comment by jaredgrady** [comment]
Deployed it. Thanks for all the help.

**Pull request merged**